### PR TITLE
MH-13308 Nibity Transcription Service

### PR DESCRIPTION
**Pull request by smarquard**
Adds support for the Nibity Transcription Service API (v1), http://nibity.com

Migrated forward from local UCT implementation for Opencast 6.x.

### CODE REVIEW

**Comment by smarquard** [comment]
For anyone who'd like to test this, I can arrange client ID and keys from Nibity.

@JamesUoM @tyfranck Stuart expressed some interest in this. It uses the same db tables as the Google Speech implementation so you should be able to backport it locally if interested.

**Comment by tyfranck** [comment]
@smarquard I want to test that, great if you can arrange client ID

**Comment by tyfranck** [comment]
@smarquard  I am currently getting a 502 error when trying to call the API. 

```
2019-10-03 10:02:24,140 | WARN  | (NibityTranscriptionService:531) - Exception when calling the captions endpoint
org.opencastproject.transcription.api.TranscriptionServiceException: Unable to submit job: API returned status 502
	at org.opencastproject.transcription.nibity.NibityTranscriptionService.createRecognitionsJob(NibityTranscriptionService.java:498) [186:opencast-transcription-service-nibity:9.0.0.SNAPSHOT]
	at org.opencastproject.transcription.nibity.NibityTranscriptionService.process(NibityTranscriptionService.java:418) [186:opencast-transcription-service-nibity:9.0.0.SNAPSHOT]
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:314) [60:opencast-common:9.0.0.SNAPSHOT]
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:273) [60:opencast-common:9.0.0.SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

**Comment by smarquard** [comment]
Currently the Nibity API only works with http or https callback media URLs (they're looking into why that is the case), so if your dev instance is running on 8080/8081/8082/etc. then please reconfigure for testing to run directly on port 80 (or put an nginx proxy in front of it and adjust the server url in custom.properties).

**Comment by smarquard** [comment]
Testing:

* Get Nibity API client ID and key 
* Configure an Opencast instance with a publicly-accessible URL, running on port 80 or 443 (so the Nibity API can request the media submission files from /transcripts/nibity/submission/...)
* Upload/ingest a demo video into Opencast using fast testing workflow (publish without editing)
* Run the Start Nibity Transcription on the event to submit to Nibity
* For testing purposes, Nibity may process the job very quickly. To start polling to retrieve results before the default time (as Nibity API doesn't support callbacks when complete), run this mysql query:

````
update oc_transcription_service_job set date_expected=now() where status='InProgress';
````

After a successful retrieval of results (polled every 60s when there are pending jobs), the Attach Nibity Transcription workflow will run. This should attach a zip file and vtt file (extracted from the zip file) to the mediapackage (inspect in asset manager). The vtt file will also be published, so the captions are available for playback (tested so far with Theodul player). The zip file will contain the .docx transcript produced by Nibity as part of the captioning process.

image

**Comment by smarquard** [comment]
Pinging @tyfranck, have you had a chance to run this through yet?

**Comment by tyfranck** [comment]
@smarquard, the requirement for opencast instance with publicly-accessible URL stopped me in my track :-( . I'm investigating few options to get one so will give it another try hopefully next week.

**Comment by tyfranck** [comment]
@smarquard , I am getting the following lines in the log which are propably errors but shows as info. Any hint? Hope my nibity client id and key are still valid.
`2019-11-28 15:14:13,353 | INFO  | (NibityTranscriptionService:514) - mp f1cd33dc-277f-4a90-8454-ebcdf16d6c15 has been submitted to nibity: file id: 5149 status 500 type mp4
2019-11-28 15:17:10,721 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5147":{"auth":504,"transcript_id":null}}
2019-11-28 15:17:11,161 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5148":{"auth":504,"transcript_id":null}}
2019-11-28 15:17:11,602 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5149":{"auth":504,"transcript_id":null}}
2019-11-28 15:18:12,164 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5147":{"auth":504,"transcript_id":null}}
2019-11-28 15:18:12,604 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5148":{"auth":504,"transcript_id":null}}
2019-11-28 15:18:13,036 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5149":{"auth":504,"transcript_id":null}}
2019-11-28 15:19:13,493 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5147":{"auth":504,"transcript_id":null}}
2019-11-28 15:19:13,925 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5148":{"auth":504,"transcript_id":null}}
2019-11-28 15:19:14,355 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5149":{"auth":504,"transcript_id":null}}
2019-11-28 15:20:14,796 | INFO  | (NibityTranscriptionService:592) - JSON from checkjob: {"5147":{"auth":504,"transcript_id":null}}`

**Comment by smarquard** [comment]
This line means a successful submission:

````
2019-11-28 15:14:13,353 | INFO | (NibityTranscriptionService:514) - mp f1cd33dc-277f-4a90-8454-ebcdf16d6c15 has been submitted to nibity: file id: 5149 status 500 type mp4
````

The 500 is a Nibity response code, not an http response code. It means "500 - Media transferred successfully"  https://documenter.getpostman.com/view/5815470/RznFpyb6?version=latest

The subsequent lines are saying that no transcript is available yet for the job ids.

**Comment by gregorydlogan** [comment]
IMO we need to translate that 500 into something sane. I would have assumed status means an error...

**Comment by tyfranck** [comment]
These are very strange definitions of status codes. A mixture of standard and their own definition: 
```
- 301 - Job ID not found
- 401 - Transcript ID not found
- 403 - Unauthorized to access transcript
- 500 - Media transferred successfully
- 502 - Unable to access media file or download corrupt
- 503 - Not authorized to access media file
- 504 - Authorized to access media file
```

The best solution is for Nibity team to stick to the standard, otherwise I agree with @gregorydlogan, we need to convert them in something sensible.

@smarquard , I uploaded a 1 min video yesterday but no transcriptions yet. What is the usual response time?

**Comment by smarquard** [comment]
On the response codes, these are not http status codes; they are defined by the API and included inside a JSON response. I could make that clearer in the logging so they're not confused with http status codes.

**Comment by gregorydlogan** [comment]
+1 for a translation layer. Looking at the full list of codes in @tyfranck's post it's doubly important. 500 vs 504 mean *very* different things, but I would assume both are errors.

**Comment by tyfranck** [comment]
@smarquard, I got the result back (1.5 days turnaround). Looks good.
I've noticed that in addition to the vtt file, nibity also provides a nibity-transcription.zip file which contains a human readable version of the caption (docx file). I think it is worth mentioning that in the documentation.

File: docs/guides/admin/docs/workflowoperationhandlers/nibity-attach-transcription-woh.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+to the media package. If the results file includes a captions file, the captions are also attached.
+
+## Parameter Table
+
+|configuration keys|description|default value|example|
+|------------------|-------|-----------|-------------|
+|transcription-job-id|This is filled out by the transcription service when starting the workflow.|EMPTY|**Should always be "${transcriptionJobId}"**|
+|target-flavor|The flavor of the caption/transcription file generated. Mandatory.|EMPTY|captions/timedtext|
+|target-tag|The tag to apply to the caption/transcription file generated. Optional.|EMPTY|archive|
+|target-caption-format|The caption format to be generated. Optional. If not entered, the raw resulting file will be attached to the media package.|EMPTY|vtt|
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/nibity-attach-transcription-woh.md:15]: What is the default if no type is provided?  VTT?  Or is that just your example?

File: modules/transcription-service-nibity/src/test/java/org/opencastproject/transcription/nibity/NibityTranscriptionServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+package org.opencastproject.transcription.nibity;
+
+import org.junit.After;
+import org.junit.Before;
+
+public class NibityTranscriptionServiceTest {
```
Comment (gregorydlogan) [inline modules/transcription-service-nibity/src/test/java/org/opencastproject/transcription/nibity/NibityTranscriptionServiceTest.java:26]: Is this empty test class deliberate?  Or are there commits missing from the squash?

File: modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/NibityAttachTranscriptionOperationHandler.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Get job id.
+    String jobId = StringUtils.trimToNull(operation.getConfiguration(TRANSCRIPTION_JOB_ID));
+    if (jobId == null)
+      throw new WorkflowOperationException(TRANSCRIPTION_JOB_ID + " missing");
+
+    // Check which tags/flavors have been configured
+    String targetFlavorOption = StringUtils.trimToNull(operation.getConfiguration(TARGET_FLAVOR));
+    String targetTagOption = StringUtils.trimToNull(operation.getConfiguration(TARGET_TAG));
+    String captionFormatOption = StringUtils.trimToNull(operation.getConfiguration(TARGET_CAPTION_FORMAT));
```
Comment (gregorydlogan) [inline modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/NibityAttachTranscriptionOperationHandler.java:108]: This variable isn't used aside from line 111, and then beyond that everything seems to assume a vtt format.  I'm fine with vtt, but it somewhat defeats the purpose of having a format configuration option when it's basically discarded...

File: modules/transcription-service-nibity/src/main/java/org/opencastproject/transcription/nibity/NibityTranscriptionService.java [hunk 4]
```diff
@@ -0,0 +1,1049 @@
...
+
+    if (!enabled) {
+      logger.info("Nibity Transcription Service disabled. If you want to enable it, please update the service configuration.");
+      return;
+    }
+
+    // Nibity API client ID and key (mandatory)
+    nibityClientId = OsgiUtil.getComponentContextProperty(cc, NIBITY_CLIENT_ID);
+    nibityClientKey = OsgiUtil.getComponentContextProperty(cc, NIBITY_CLIENT_KEY);
+    logger.info("Nibity Transcription Service enabled with client id {}", nibityClientId);
```
Comment (gregorydlogan) [inline modules/transcription-service-nibity/src/main/java/org/opencastproject/transcription/nibity/NibityTranscriptionService.java:206]: We should be validating that these aren't blank at the very least IMO

File: modules/transcription-service-nibity/src/main/java/org/opencastproject/transcription/nibity/NibityTranscriptionService.java [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+   * Called by process(Job job)
+   */
+  void createRecognitionsJob(String mpId, Track track, String languageCode) throws TranscriptionServiceException, IOException {
+
+    String filename = addMediaFileToLocalStorage(mpId, track);
+    String mediaUrl = serverUrl + SUBMISSION_PATH + filename;
+
+    logger.info("Media URL in intermediate storage: {}", mediaUrl);
+
+    CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
```
Comment (gregorydlogan) [inline modules/transcription-service-nibity/src/main/java/org/opencastproject/transcription/nibity/NibityTranscriptionService.java:440]: I'd use `makeHttpClient` here, adding a paramter for the `RequestConfig` to the method

**Review by gregorydlogan — CHANGES_REQUESTED** [review]
Minor nits, but looks good to me otherwise.

**Comment by gregorydlogan** [comment]
@smarquard is this going to move forward?

**Comment by smarquard** [comment]
It is indeed. I will make sure it's wrapped up in time for the 9.x branch.

**Comment by KatrinIhler** [comment]
@smarquard Conflicts!

**Comment by gregorydlogan** [comment]
@smarquard any progress here?

**Comment by gregorydlogan** [comment]
Reminder: As a feature this has until 2020-10-05 to be merged for it to make it in 9.0!

**Comment by KatrinIhler** [comment]
@smarquard What's the status here? If you're planning to get this into OC 10, that deadline is approaching. If not, it might make more sense to close this for now and re-open when it's ready.

**Comment by KatrinIhler** [comment]
As per our recent discussion in the dev meeting and the proposal that followed, I'm going to close this PR since this has seen no updates within the past 1 year. This does not constitute a rejection, feel free to re-open once you have time to address the review findings. :)

**Comment by KatrinIhler** [comment]
As per our recent discussion in the dev meeting and the proposal that followed, I'm going to close this PR since this has seen no updates within the past 1 year. This does not constitute a rejection, feel free to re-open once you have time to address the review findings. :)

**Comment by smarquard** [comment]
We're still running this in production at UCT but haven't had time to resolve the concerns for upstream contribution due to local COVID-19 priorities. We aim to get back to it soon.

**Pull request closed without merging**